crates/ruff_python_formatter/src/string/normalize.rs

```diff
@@ -127,22 +68,110 @@ impl StringNormalizer {
             }
         } else {
             self.quoting
-        };
+        }
+    }
 
-        match quoting {
+    /// Computes the strings preferred quotes.
+    pub(crate) fn choose_quotes(&self, string: &StringPart, locator: &Locator) -> QuoteSelection {
+        let raw_content = locator.slice(string.content_range());
+        let first_quote_or_normalized_char_offset = raw_content
+            .bytes()
+            .position(|b| matches!(b, b'\\' | b'"' | b'\'' | b'\r' | b'{'));
```

Comment:
If it's not an f-string, you can omit `{`, I think? Similarly, if it's a raw string, you can omit `\\`... (In that case, you could actually use `memchr3`, but perhaps not worth it.)

Reply:
I considered special casing but decided against it because I want to avoid the extra complexity and it doesn't have to be perfect, for as long as it avoids the "expensive" normalization for most strings. 

 Using `memchr` would be nice... but `normalize` is already now almost gone from the benchmark profiles. So I consider it "good enough"